drltrace/drltrace_frontend.cpp

```diff
@@ -0,0 +1,314 @@
+
+    if (drfront_access(absolute_logdir_path, DRFRONT_WRITE, &result) != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("cannot find logdir %s", absolute_logdir_path);
+    if (!result)
+        DRLTRACE_ERROR("cannot write log file into %s", absolute_logdir_path);
+}
+
+int
+_tmain(int argc, const TCHAR *targv[])
+{
+    char drlibpath[MAXIMUM_PATH];
+#ifdef WINDOWS
+    static const char *libname = "drltracelib.dll";
+#else
+    static const char *libname = "libdrltracelib.so";
+#endif
+#ifdef DEBUG
+    sprintf(drlibpath, "debug/%s", libname);
+#else
+    sprintf(drlibpath, "release/%s", libname);
+#endif
+    void *inject_data;
+    int exitcode;
+    char **argv;
+    char *tmp;
+    const char *target_app_name;
+
+    char full_target_app_path[MAXIMUM_PATH];
+    char tmp_path[MAXIMUM_PATH];
+    char full_frontend_path[MAXIMUM_PATH];
+    char full_dr_root_path[MAXIMUM_PATH];
+    char full_drlibtrace_path[MAXIMUM_PATH];
+
+    int last_index;
+    std::string parse_err;
+    drfront_status_t sc;
+
+#if defined(WINDOWS) && !defined(_UNICODE)
+# error _UNICODE must be defined
+#else
+    /* Convert to UTF-8 if necessary */
+    sc = drfront_convert_args((const TCHAR **)targv, &argv, argc);
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("failed to process args, error code = %d\n", sc);
+#endif
+
+    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_FRONTEND, argc, (const char **)argv,
+                                       &parse_err, &last_index) || argc < 2)
+        DRLTRACE_ERROR("Usage error: %s\n Usage:\n%s\n", parse_err.c_str(),
+              droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
+
+    if (op_version.get_value()) {
+        print_version();
+        return 0;
+    }
+    if (op_help.get_value()) {
+        printf("Usage:\n%s", droption_parser_t::usage_long(DROPTION_SCOPE_ALL).c_str());
+        return 0;
+    }
+
+    target_app_name = argv[last_index];
+    if (target_app_name == NULL) {
+        DRLTRACE_ERROR("Usage error, target application is not specified.\n Usage:\n%s\n",
+                       droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
+    }
+    sc = drfront_get_app_full_path(target_app_name, full_target_app_path,
+                                   BUFFER_SIZE_ELEMENTS(full_target_app_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_app_full_path failed on %s, error code = %d\n",
+                       target_app_name, sc);
+
+    /* get DR's root directory and drltracelib.dll full path */
+    sc = drfront_get_app_full_path(argv[0], full_frontend_path,
+                                   BUFFER_SIZE_ELEMENTS(full_frontend_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_app_full_path failed on %s, error code = %d\n",
+                       argv[0], sc);
+
+    tmp = full_frontend_path + strlen(full_frontend_path) - 1;
+
+    while (*tmp != DIRSEP && *tmp != ALT_DIRSEP && tmp > full_frontend_path)
```

Comment:
suggest: comment stating assumptions: you assume this binary is in some subdir like bin/?